### DISCUSSION TITLE
Question: Does the regex-syntax crate optimize its HIR?

### DISCUSSION THREAD

**Discussion by ckaran** [rust-lang/regex#915] [Q&A] [answered]
## Background

I'm working on writing yet another fuzzing framework (YAFF for short), but I'm doing it via an interesting twist.  I'm using regular expressions (specifically, what [`regex-syntax`](https://crates.io/crates/regex-syntax) supports) to describe the space of byte strings that a type suggests a fuzzer should explore.  Roughly speaking, the idea is the following:

1. A type implements both [`Arbitrary`](https://crates.io/crates/arbitrary) and my own trait `Semiarbitrary`.
2. `Semiarbitrary` has an associated function defined that returns a `&str`, which is a regular expression that obeys the [syntax of the `regex` crate](https://docs.rs/regex/latest/regex/#syntax).
3. I parse this expression using the [`regex-syntax`](https://crates.io/crates/regex-syntax) crate, and I create strings that would match the regular expression given in step 2 (this is the fuzzing part).
4. These strings are used to create new instances of the type being tested via the type's implementation of [`Arbitrary`](https://crates.io/crates/arbitrary).  
    1. If  [`Arbitrary`](https://crates.io/crates/arbitrary) returns an error of type [`IncorrectFormat`](https://docs.rs/arbitrary/latest/arbitrary/enum.Error.html#variant.IncorrectFormat), then I know that the original regex matched more strings than can be accepted.  I start exploring the space around the rejected string to see if can improve the regular expression by subtracting out chunks of space that the type's implementation of [`Arbitrary`](https://crates.io/crates/arbitrary) will reject with an [`IncorrectFormat`](https://docs.rs/arbitrary/latest/arbitrary/enum.Error.html#variant.IncorrectFormat) error.
    2. If I get a valid value, but the test fails I log the string in my fuzzing database and explore/reduce it to find the minimal regex that produces test failures.
    3. If the test passes, I try to find a point in byte-string space that is 'half-way' to another point, and *if it can be matched by the type's original regex* I run it through the tests.  Otherwise I look for another point in space.
5. Repeat the above steps until I'm stopped, or I've found lots of bugs.

## The Issue

Fuzzing rarely produces truly optimal output, and part of that problem is that it can generate regular expressions that are horribly inefficient.  E.g., alternations that repeatedly sample the same region of byte string space, or have partitioned space into a billion small points which could all be joined into a much larger combined space.  I want to know the following:

1. Does the [`regex-syntax`](https://crates.io/crates/regex-syntax) crate optimize its [`Hir`](https://docs.rs/regex-syntax/latest/regex_syntax/hir/struct.Hir.html)?
2. Is there a way of re-optimizing the [`Hir`](https://docs.rs/regex-syntax/latest/regex_syntax/hir/struct.Hir.html) after mutating it?

Ideally for me, the answer to both questions would be 'yes.'  That way I can mutate to my heart's content, but still have a way of reducing the regular expression down to the smallest subspace that a) contains valid instances only, and b) still causes a test failure.  

I hope all of this makes sense.

**Comment by BurntSushi** [comment, marked as answer]
> Does the [regex-syntax](https://crates.io/crates/regex-syntax) crate optimize its [Hir](https://docs.rs/regex-syntax/latest/regex_syntax/hir/struct.Hir.html)?

It does a little bit of optimization today. But very little. Here's one example:

https://github.com/rust-lang/regex/blob/4075d39345dea65d3367ace7cea2c248ae9ce11a/regex-syntax/src/hir/mod.rs#L398-L402

I do have a rather large WIP patchset (that will become `regex-syntax 0.8` at some point) that beefs up the optimizations. Example 1:

https://github.com/rust-lang/regex/blob/7ecd9ff770ff7891bf3b88c5f65596d542658c88/regex-syntax/src/hir/mod.rs#L342-L405

Example 2:

https://github.com/rust-lang/regex/blob/7ecd9ff770ff7891bf3b88c5f65596d542658c88/regex-syntax/src/hir/mod.rs#L410-L460

But... It looks like you're asking about _particular_ optimizations. For example:

> alternations that repeatedly sample the same region of byte string space, or have partitioned space into a billion small points which could all be joined into a much larger combined space

I actually need to see examples of this to be honest. I bet I could make a good guess at what you're saying. I would say that `regex-syntax` does not do this kind of optimization today. I _suspect_ it might not ever. I have [some optimizations like this in another WIP branch for `regex-automata`](https://github.com/BurntSushi/regex-automata/blob/3900ee6610873ae4deef6b5913ff8ff0dec4755f/src/nfa/thompson/literal_trie.rs), but those are being applied while compiling an NFA. I don't have any plans to port those to the HIR. The principle reason is that once you get into things like "reducing the size of alternations based on literals," things get quite complex and there has to be a balance between "produce the simplest possible HIR" and "don't take too much time building an HIR."

There are other complexities. For example, `regex-syntax` doesn't know anything about _match semantics_. In some regex engines, `sam|samwise` will match `sam` in `samwise`, and in other regex engines, it will match `samwise`. The match semantics of the former are what the `regex` crate implements, and they directly imply that the order of literals in an alternation matters. So for example, if you have `zap|z|zapper`, you might think an equivalent regex is `z(ap(per)?)?`, but it actually isn't, because that regex will match `zapper` in `zapper` when it should have matched `zap` when using the `regex` crate. So the simplified regex in that case would be `z(ap||apper)`.

So I think in order to really consider your question, it would be best to provide specifics. :-)

> Is there a way of re-optimizing the [Hir](https://docs.rs/regex-syntax/latest/regex_syntax/hir/struct.Hir.html) after mutating it?

Can you show me your code that mutates an `Hir`? :-) I contend that such things are impossible, and that is by design. The only thing you can do is structural induction over the `Hir` and produce a _new_ `Hir`. And the only way to produce an `Hir` is through its "smart" constructors. And those "smart" constructors are what is responsible for doing the optimizations. This makes it, for example, impossible to build an `Hir` with an alternation containing fewer than 2 sub-expressions. You can't even remove those sub-expressions after the `Hir` is built because there are no mutating APIs.

**Reply by ckaran** [reply]
> I actually need to see examples of this to be honest. I bet I could make a good guess at what you're saying.

`(a|b|c|d|e)` could be simplified to `[a-e]`.  `([a-y]|[b-z])` could be simplified to `[a-z]`.  I'm not after full, perfect optimizations, more like anything that helps reduce stuff

> Can you show me your code that mutates an Hir? 

Sorry, you're right, you cannot mutate it *in-place*, so my 'mutations' are always new `Hir` that are 'mutations' of the original.  I got sloppy with my language, which is a **bad** thing to do when dealing with the author of a regex crate! :sweat_smile:

**Reply by BurntSushi** [reply]
> (a|b|c|d|e) could be simplified to [a-e]. ([a-y]|[b-z]) could be simplified to [a-z]. I'm not after full, perfect optimizations, more like anything that helps reduce stuff

Both of those will be in `regex-syntax 0.8`. Actually, those are exactly the two optimizations present in my "Example 2" above. :-)

> Sorry, you're right, you cannot mutate it in-place, so my 'mutations' are always new Hir that are 'mutations' of the original. I got sloppy with my language, which is a bad thing to do when dealing with the author of a regex crate! sweat_smile

Ah I see yes. In that case, the answer is yes, the optimizations will be applied automatically. I specifically made the optimizations part of the construction of `Hir` itself, so that anyone using it, no matter how they build it, benefits from them. There isn't a separate "optimization pass" on HIR.

**Reply by ckaran** [reply]
Thank you for answering my questions, and thank you for making this crate!  I'm still hacking around on my own crate, but `regex-syntax` has already made my life **much** easier!
